### PR TITLE
[AF-1019]: Workspace and Repository service can rename project or repository if that name already exist

### PR DESCRIPTION
**Pull request by adrielparedes**
@paulovmr @ederign Can you review this please?

related to:
https://github.com/kiegroup/kie-wb-common/pull/1459

### CODE REVIEW

**Comment by paulovmr** [comment]
@adrielparedes Looks good. Is there a PR for kie-wb-common changing `ExampleServiceImpl` to call these new methods and/or removing their own?

**Comment by adrielparedes** [comment]
@paulovmr you know, I forgot about that part, let me see how to remove them

**Comment by adrielparedes** [comment]
jenkins retest this

**Comment by adrielparedes** [comment]
jenkins retest this

File: uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/service/WorkspaceProjectService.java [hunk 1]
```diff
@@ -44,6 +44,9 @@ WorkspaceProject newProject(final OrganizationalUnit organizationalUnit,
                                 final POM pom,
                                 final DeploymentMode mode);
 
+    String createProjectName(OrganizationalUnit organizationalUnit,
+                             String name);
```
Comment (jomarko) [inline uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/service/WorkspaceProjectService.java]: Could we please use `final OrganizationalUnit organizationalUnit` and `final String name`?

File: uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/WorkspaceProjectServiceImpl.java [hunk 2]
```diff
@@ -159,6 +164,19 @@ public WorkspaceProject newProject(final OrganizationalUnit organizationalUnit,
         return workspaceProject;
     }
 
+    @Override
+    public String createProjectName(OrganizationalUnit organizationalUnit,
+                                    String name) {
```
Comment (jomarko) [inline uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/WorkspaceProjectServiceImpl.java]: Same comment as in the interface.

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/RepositoryServiceImpl.java [hunk 3]
```diff
@@ -280,6 +289,18 @@ public Repository createRepository(final OrganizationalUnit organizationalUnit,
         }
     }
 
+    protected String createNewAliasIfNecessary(String alias,
+                                             Space space) {
```
Comment (jomarko) [inline uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/RepositoryServiceImpl.java]: Formatting , also could we use `final String alias` and `final Space space`?

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/RepositoryServiceImpl.java [hunk 4]
```diff
@@ -293,17 +314,21 @@ protected ConfigGroup findRepositoryConfig(final String alias) {
     }
 
     @Override
-    public void removeRepository(Space space, String alias) {
+    public void removeRepository(Space space,
+                                 String alias) {
```
Comment (jomarko) [inline uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/RepositoryServiceImpl.java]: Similar comment as in the interface, could we use `final`?

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/RepositoryServiceImpl.java [hunk 5]
```diff
@@ -206,11 +209,13 @@ public Repository getRepositoryFromSpace(final Space space,
...
         Space space = spacesAPI.resolveSpace(root.toURI()).orElseThrow(() -> new IllegalArgumentException("Cannot resolve space from given path: " + root));
-        return configuredRepositories.getRepositoryByRootPath(space, root);
+        return configuredRepositories.getRepositoryByRootPath(space,
+                                                              root);
     }
 
     @Override
-    public Repository getRepository(Space space, Path root) {
+    public Repository getRepository(Space space,
+                                    Path root) {
```
Comment (jomarko) [inline uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/RepositoryServiceImpl.java]: Could we please use `final Space space` and `final Path root`?

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/RepositoryServiceImpl.java [hunk 6]
```diff
@@ -449,7 +482,8 @@ private Repository createRepository(final String scheme,
         }
     }
 
-    private Repository createRepository(final ConfigGroup repositoryConfig, Space space) {
+    private Repository createRepository(final ConfigGroup repositoryConfig,
+                                        Space space) {
```
Comment (jomarko) [inline uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/RepositoryServiceImpl.java]: Could we please use `final Space space`?

**Comment by jhrcek** [comment]
Jenkins execute full downstream build

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/RepositoryServiceImpl.java [hunk 7]
```diff
@@ -314,17 +339,22 @@ public void removeRepository(Space space, String alias) {
     }
 
     @Override
-    public void removeRepositories(Space space, Set<String> aliases) {
+    public void removeRepositories(Space space,
+                                   Set<String> aliases) {
```
Comment (jomarko) [inline uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/RepositoryServiceImpl.java]: Could we please use `final Space space` and `final Set<String> aliases`?

**Comment by adrielparedes** [comment]
Jenkins execute full downstream build

**Comment by jhrcek** [comment]
@manstis the test `org.kie.workbench.common.services.backend.compiler.nio.kie.KieDefaultMavenCompilerTest` failed, with `Failed to open server socket for /127.0.0.1:9418`, because  `integration-test (default) @ kie-wb-common-ala-distribution` was running at the same time, which started wildfly swarm that occupied this port. What would you suggest to prevent this kind of random failures in the future?

1. make the `kie-wb-common-ala-distribution` tests start wildfly swarm with port offset?
2. use [a hack like this](https://github.com/kiegroup/appformer/blob/master/uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-jgit/src/test/java/org/uberfire/java/nio/fs/jgit/AbstractTestInfra.java#L182-L194) in the `KieDefaultMavenCompilerTest` and its friends to avoid port conflicts?

<details>
  <summary>build output showing interleaving</summary>
  <pre>
[INFO] --- wildfly-swarm-plugin:2017.8.0:start (start) @ kie-wb-common-ala-distribution ---
[INFO] Starting .war
[INFO] Scanning for needed WildFly Swarm fractions with mode: when_missing
[INFO] Detected fractions: 
[INFO] Using fractions: 
Wed Feb 28 10:33:38 EST 2018 INFO [org.wildfly.swarm.bootstrap] (main) Dependencies not bundled; resolving from M2REPO.
2018-02-28 10:33:38,710 INFO  [org.wildfly.swarm] (main) WFSWARM0013: Installed fraction:                  Logging - STABLE          org.wildfly.swarm:logging:2017.8.0
2018-02-28 10:33:38,720 INFO  [org.wildfly.swarm] (main) WFSWARM0013: Installed fraction:          Bean Validation - STABLE          org.wildfly.swarm:bean-validation:2017.8.0
2018-02-28 10:33:38,720 INFO  [org.wildfly.swarm] (main) WFSWARM0013: Installed fraction:        CDI Configuration - STABLE          org.wildfly.swarm:cdi-config:2017.8.0
2018-02-28 10:33:38,720 INFO  [org.wildfly.swarm] (main) WFSWARM0013: Installed fraction:             Transactions - STABLE          org.wildfly.swarm:transactions:2017.8.0
2018-02-28 10:33:38,720 INFO  [org.wildfly.swarm] (main) WFSWARM0013: Installed fraction:                      CDI - STABLE          org.wildfly.swarm:cdi:2017.8.0
2018-02-28 10:33:38,721 INFO  [org.wildfly.swarm] (main) WFSWARM0013: Installed fraction:                 Undertow - STABLE          org.wildfly.swarm:undertow:2017.8.0
2018-02-28 10:33:38,721 INFO  [org.wildfly.swarm] (main) WFSWARM0013: Installed fraction:                   JAX-RS - STABLE          org.wildfly.swarm:jaxrs:2017.8.0
2018-02-28 10:33:40,288 INFO  [org.jboss.msc] (main) JBoss MSC version 1.2.6.Final
2018-02-28 10:33:40,349 INFO  [org.jboss.as] (MSC service thread 1-6) WFLYSRV0049: WildFly Swarm 2017.8.0 (WildFly Core 2.2.1.Final) starting
2018-02-28 10:33:40,390 INFO  [org.wildfly.swarm] (MSC service thread 1-6) WFSWARM0019: Install MSC service for command line args: []
2018-02-28 10:33:41,014 INFO  [org.wildfly.extension.io] (ServerService Thread Pool -- 13) WFLYIO001: Worker 'default' has auto-configured to 16 core threads with 128 task threads based on your 8 available processors
2018-02-28 10:33:41,030 WARN  [org.jboss.as.txn] (ServerService Thread Pool -- 16) WFLYTX0013: Node identifier property is set to the default value. Please make sure it is unique.
2018-02-28 10:33:41,033 INFO  [org.jboss.as.naming] (ServerService Thread Pool -- 17) WFLYNAM0001: Activating Naming Subsystem
2018-02-28 10:33:41,069 INFO  [org.jboss.as.naming] (MSC service thread 1-1) WFLYNAM0003: Starting Naming Service
2018-02-28 10:33:41,072 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-4) WFLYUT0003: Undertow 1.4.11.Final starting
2018-02-28 10:33:41,112 INFO  [org.jboss.as.security] (ServerService Thread Pool -- 22) WFLYSEC0002: Activating Security Subsystem
2018-02-28 10:33:41,118 INFO  [org.jboss.as.security] (MSC service thread 1-7) WFLYSEC0001: Current PicketBox version=4.9.6.Final
2018-02-28 10:33:41,157 INFO  [org.xnio] (MSC service thread 1-6) XNIO version 3.4.3.Final
2018-02-28 10:33:41,170 INFO  [org.xnio.nio] (MSC service thread 1-6) XNIO NIO Implementation Version 3.4.3.Final
2018-02-28 10:33:41,341 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-6) WFLYUT0012: Started server default-server.
2018-02-28 10:33:41,386 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-6) WFLYUT0006: Undertow HTTP listener default listening on [0:0:0:0:0:0:0:0]:8080
2018-02-28 10:33:41,503 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0025: WildFly Swarm 2017.8.0 (WildFly Core 2.2.1.Final) started in 1244ms - Started 88 of 97 services (18 services are lazy, passive or on-demand)
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 11.204 sec - in org.kie.workbench.common.services.backend.compiler.nio.kie.KieDefaultMavenIncrementalCompilerTest
Running org.kie.workbench.common.services.backend.compiler.nio.kie.KieMavenOutputTest
2018-02-28 10:33:45,788 INFO  [org.wildfly.swarm.runtime.deployer] (main) deploying kie-wb-common-ala-distribution-7.7.0-SNAPSHOT.war
2018-02-28 10:33:45,817 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-3) WFLYSRV0027: Starting deployment of "kie-wb-common-ala-distribution-7.7.0-SNAPSHOT.war" (runtime-name: "kie-wb-common-ala-distribution-7.7.0-SNAPSHOT.war")
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.88 sec - in org.kie.workbench.common.services.backend.compiler.nio.kie.KieMavenOutputTest
Running org.kie.workbench.common.services.backend.compiler.nio.kie.KieDefaultMavenCompilerTest
2018-02-28 10:33:51,124 WARN  [org.jboss.as.server.deployment] (MSC service thread 1-8) WFLYSRV0059: Class Path entry jaxb-core.jar in /content/kie-wb-common-ala-distribution-7.7.0-SNAPSHOT.war/WEB-INF/lib/jaxb-impl-2.2.11.jar  does not point to a valid jar for a Class-Path reference.
2018-02-28 10:33:51,161 WARN  [org.jboss.as.server.deployment] (MSC service thread 1-8) WFLYSRV0059: Class Path entry jaxb-api.jar in /content/kie-wb-common-ala-distribution-7.7.0-SNAPSHOT.war/WEB-INF/lib/jaxb-core-2.2.11.jar  does not point to a valid jar for a Class-Path reference.
2018-02-28 10:33:51,740 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-5) WFLYSRV0018: Deployment "deployment.kie-wb-common-ala-distribution-7.7.0-SNAPSHOT.war" is using a private module ("org.jboss.jts:main") which may be changed or removed in future versions without notice.
2018-02-28 10:33:51,780 INFO  [org.jboss.weld.deployer] (MSC service thread 1-4) WFLYWELD0003: Processing weld deployment kie-wb-common-ala-distribution-7.7.0-SNAPSHOT.war
2018-02-28 10:33:51,826 INFO  [org.hibernate.validator.internal.util.Version] (MSC service thread 1-4) HV000001: Hibernate Validator 5.2.4.Final
2018-02-28 10:33:52,089 INFO  [org.jboss.weld.Version] (MSC service thread 1-5) WELD-000900: 2.3.5 (Final)
2018-02-28 10:33:52,111 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-6) WFLYUT0018: Host default-host starting
2018-02-28 10:33:53,641 INFO  [org.jboss.resteasy.resteasy_jaxrs.i18n] (ServerService Thread Pool -- 3) RESTEASY002225: Deploying javax.ws.rs.core.Application: class org.guvnor.ala.services.swarm.App$Proxy$_$$_WeldClientProxy
2018-02-28 10:33:53,671 WARN  [org.jboss.as.weld] (ServerService Thread Pool -- 3) WFLYWELD0052: Using deployment classloader to load proxy classes for module com.fasterxml.jackson.jaxrs.jackson-jaxrs-json-provider:main. Package-private access will not work. To fix this the module should declare dependencies on [org.jboss.weld.core, org.jboss.weld.spi]
2018-02-28 10:33:53,706 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 3) WFLYUT0021: Registered web context: /
2018-02-28 10:33:53,741 INFO  [org.jboss.as.server] (main) WFLYSRV0010: Deployed "kie-wb-common-ala-distribution-7.7.0-SNAPSHOT.war" (runtime-name : "kie-wb-common-ala-distribution-7.7.0-SNAPSHOT.war")
2018-02-28 10:33:53,743 INFO  [org.wildfly.swarm] (main) WFSWARM99999: WildFly Swarm is Ready
[INFO] 
[INFO] --- maven-failsafe-plugin:2.18.1:integration-test (default) @ kie-wb-common-ala-distribution ---
[INFO] Failsafe report directory: /home/jenkins/workspace/appformer-downstream-pullrequests/downstream-repos/kie-wb-common/kie-wb-common-ala/kie-wb-common-ala-distribution/target/failsafe-reports

-------------------------------------------------------
 T E S T S
-------------------------------------------------------

Results :

Tests run: 0, Failures: 0, Errors: 0, Skipped: 0

[INFO] 
[INFO] --- wildfly-swarm-plugin:2017.8.0:stop (stop) @ kie-wb-common-ala-distribution ---
[INFO] 
[INFO] --- maven-failsafe-plugin:2.18.1:verify (default) @ kie-wb-common-ala-distribution ---
[INFO] Failsafe report directory: /home/jenkins/workspace/appformer-downstream-pullrequests/downstream-repos/kie-wb-common/kie-wb-common-ala/kie-wb-common-ala-distribution/target/failsafe-reports
[INFO] 
[INFO] --- maven-install-plugin:2.5.2:install (default-install) @ kie-wb-common-ala-distribution ---
[INFO] Installing /home/jenkins/workspace/appformer-downstream-pullrequests/downstream-repos/kie-wb-common/kie-wb-common-ala/kie-wb-common-ala-distribution/target/kie-wb-common-ala-distribution-7.7.0-SNAPSHOT.war to /home/jenkins/.m2/repository/org/kie/workbench/kie-wb-common-ala-distribution/7.7.0-SNAPSHOT/kie-wb-common-ala-distribution-7.7.0-SNAPSHOT.war
[INFO] Installing /home/jenkins/workspace/appformer-downstream-pullrequests/downstream-repos/kie-wb-common/kie-wb-common-ala/kie-wb-common-ala-distribution/pom.xml to /home/jenkins/.m2/repository/org/kie/workbench/kie-wb-common-ala-distribution/7.7.0-SNAPSHOT/kie-wb-common-ala-distribution-7.7.0-SNAPSHOT.pom
[INFO] Installing /home/jenkins/workspace/appformer-downstream-pullrequests/downstream-repos/kie-wb-common/kie-wb-common-ala/kie-wb-common-ala-distribution/target/kie-wb-common-ala-distribution-7.7.0-SNAPSHOT-tests.jar to /home/jenkins/.m2/repository/org/kie/workbench/kie-wb-common-ala-distribution/7.7.0-SNAPSHOT/kie-wb-common-ala-distribution-7.7.0-SNAPSHOT-tests.jar
[INFO] Installing /home/jenkins/workspace/appformer-downstream-pullrequests/downstream-repos/kie-wb-common/kie-wb-common-ala/kie-wb-common-ala-distribution/target/kie-wb-common-ala-distribution-7.7.0-SNAPSHOT-sources.jar to /home/jenkins/.m2/repository/org/kie/workbench/kie-wb-common-ala-distribution/7.7.0-SNAPSHOT/kie-wb-common-ala-distribution-7.7.0-SNAPSHOT-sources.jar
[INFO] Installing /home/jenkins/workspace/appformer-downstream-pullrequests/downstream-repos/kie-wb-common/kie-wb-common-ala/kie-wb-common-ala-distribution/target/kie-wb-common-ala-distribution-7.7.0-SNAPSHOT-test-sources.jar to /home/jenkins/.m2/repository/org/kie/workbench/kie-wb-common-ala-distribution/7.7.0-SNAPSHOT/kie-wb-common-ala-distribution-7.7.0-SNAPSHOT-test-sources.jar
[INFO] Installing /home/jenkins/workspace/appformer-downstream-pullrequests/downstream-repos/kie-wb-common/kie-wb-common-ala/kie-wb-common-ala-distribution/target/kie-wb-common-ala-distribution-7.7.0-SNAPSHOT-javadoc.jar to /home/jenkins/.m2/repository/org/kie/workbench/kie-wb-common-ala-distribution/7.7.0-SNAPSHOT/kie-wb-common-ala-distribution-7.7.0-SNAPSHOT-javadoc.jar
[INFO] Installing /home/jenkins/workspace/appformer-downstream-pullrequests/downstream-repos/kie-wb-common/kie-wb-common-ala/kie-wb-common-ala-distribution/target/kie-wb-common-ala-distribution-7.7.0-SNAPSHOT-swarm.jar to /home/jenkins/.m2/repository/org/kie/workbench/kie-wb-common-ala-distribution/7.7.0-SNAPSHOT/kie-wb-common-ala-distribution-7.7.0-SNAPSHOT-swarm.jar
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building Kie Workbench - Common - Forms - ADF - Engine - API 7.7.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:3.0.0:clean (default-clean) @ kie-wb-common-forms-adf-engine-api ---
[INFO] Deleting /home/jenkins/workspace/appformer-downstream-pullrequests/downstream-repos/kie-wb-common/kie-wb-common-forms/kie-wb-common-forms-core/kie-wb-common-forms-adf/kie-wb-common-forms-adf-engine/kie-wb-common-forms-adf-engine-api (includes = [repository/**, .niogit/**, .index/**, .security/**], excludes = [])
[INFO] 
[INFO] --- maven-enforcer-plugin:3.0.0-M1:enforce (enforce-plugin-versions) @ kie-wb-common-forms-adf-engine-api ---
[INFO] 
[INFO] --- maven-enforcer-plugin:3.0.0-M1:enforce (ban-blacklisted-dependencies) @ kie-wb-common-forms-adf-engine-api ---
[INFO] 
[INFO] --- maven-enforcer-plugin:3.0.0-M1:enforce (ban-duplicated-classes) @ kie-wb-common-forms-adf-engine-api ---
Tests run: 4, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 10.05 sec <<< FAILURE! - in org.kie.workbench.common.services.backend.compiler.nio.kie.KieDefaultMavenCompilerTest
buildWithPullRebaseUberfireTest(org.kie.workbench.common.services.backend.compiler.nio.kie.KieDefaultMavenCompilerTest)  Time elapsed: 0.03 sec  <<< ERROR!
org.uberfire.java.nio.IOException: java.io.IOException: Failed to open server socket for /127.0.0.1:9418
	at java.net.PlainSocketImpl.socketBind(Native Method)
	at java.net.AbstractPlainSocketImpl.bind(AbstractPlainSocketImpl.java:387)
	at java.net.ServerSocket.bind(ServerSocket.java:375)
	at java.net.ServerSocket.<init>(ServerSocket.java:237)
	at org.uberfire.java.nio.fs.jgit.daemon.git.Daemon.start(Daemon.java:295)
	at org.uberfire.java.nio.fs.jgit.JGitFileSystemProvider.buildAndStartDaemon(JGitFileSystemProvider.java:474)
	at org.uberfire.java.nio.fs.jgit.JGitFileSystemProvider.newFileSystem(JGitFileSystemProvider.java:596)
	at org.uberfire.java.nio.file.FileSystems.newFileSystem(FileSystems.java:122)
	at org.uberfire.java.nio.file.FileSystems.newFileSystem(FileSystems.java:78)
	at org.uberfire.io.impl.AbstractIOService.newFileSystem(AbstractIOService.java:231)
	at org.uberfire.mocks.FileSystemTestingUtils.setupJGitRepository(FileSystemTestingUtils.java:76)
	at org.uberfire.mocks.FileSystemTestingUtils.setup(FileSystemTestingUtils.java:48)
	at org.uberfire.mocks.FileSystemTestingUtils.setup(FileSystemTestingUtils.java:41)
	at org.kie.workbench.common.services.backend.compiler.nio.kie.KieDefaultMavenCompilerTest.setUp(KieDefaultMavenCompilerTest.java:65)
  </pre>
</details>

**Comment by manstis** [comment]
@jhrcek Either is better than we have now. Port offsets are fine.. until we later have other tests that also use (the same) port offset.. so I tend to think the "hack" is actually the better option!

**Comment by jhrcek** [comment]
@manstis ok, I will look into using the option 2 in all the affected tests soon.

**Pull request merged**